core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala

```diff
@@ -0,0 +1,338 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.stats
+
+import java.util.Objects
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.delta._
+import org.apache.spark.sql.delta.actions.AddFile
+import org.apache.spark.sql.delta.files.{TahoeFileIndex, TahoeLogFileIndex}
+import org.apache.spark.sql.delta.metering.DeltaLogging
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.planning.PhysicalOperation
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.types.StructType
+
+/**
+ * Before query planning, we prepare any scans over delta tables by pushing
+ * any projections or filters in allowing us to gather more accurate statistics
+ * for CBO and metering.
+ *
+ * Note the following
+ * - This rule also ensures that all reads from the same delta log use the same snapshot of log
+ *   thus providing snapshot isolation.
+ * - If this rule is invoked within an active [[OptimisticTransaction]], then the scans are
+ *   generated using the transaction.
+ */
+trait PrepareDeltaScanBase extends Rule[LogicalPlan]
+  with PredicateHelper
+  with DeltaLogging { self: PrepareDeltaScan =>
+
+  private val snapshotIsolationEnabled = spark.conf.get(DeltaSQLConf.DELTA_SNAPSHOT_ISOLATION)
+
+  /**
+   * Tracks the first-access snapshots of other logs planned by this rule. The snapshots are
+   * the keyed by the log's unique id. Note that the lifetime of this rule is a single
+   * query, therefore, the map tracks the snapshots only within a query.
+   */
+  private val scannedSnapshots =
+    new java.util.concurrent.ConcurrentHashMap[(String, Path), Snapshot]
+
+  /**
+   * Gets the [[DeltaScanGenerator]] for the given log, which will be used to generate
+   * [[DeltaScan]]s. Every time this method is called on a log within the lifetime of this
+   * rule (i.e., the lifetime of the query for which this rule was instantiated), the returned
+   * generator will read a snapshot that is pinned on the first access for that log.
+   *
+   * Internally, it will use the snapshot of the file index, the snapshot of the active transaction
+   * (if any), or the latest snapshot of the given log.
+   */
+  protected def getDeltaScanGenerator(index: TahoeLogFileIndex): DeltaScanGenerator = {
```

Comment:
this looks very similar to the existing code ActiveOptTxnRule. ... again.. elaborate what the difference is.